pilot/pkg/bootstrap/server.go

```diff
@@ -206,18 +243,50 @@ func (s *Server) Start(stop <-chan struct{}) error {
 		}
 	}
 
+	// grpcServer is shared by Galley, CA, XDS - must Serve at the end, but before 'wait'
+	go func() {
+		if err := s.grpcServer.Serve(s.grpcListener); err != nil {
+			log.Warna(err)
+		}
+	}()
+
+	if !s.waitForCacheSync(stop) {
+		return fmt.Errorf("failed to sync cache")
+	}
+	log.Infof("starting discovery service at http=%s grpc=%s", s.HttpListener.Addr(),
+		s.grpcListener.Addr())
+
+	// At this point we are ready
+	go func() {
+		if err := s.httpServer.Serve(s.HttpListener); err != nil {
```

Comment:
why split insecure http and grpc server start from `initDiscoveryService`, but keeps secure server start there.  I think every component should register a start call back.

Reply:
I don't think start callback is required - for some things it's cleaner to have a more explicit ordering. 

The secure GRPC depends on the certificates, and should be loaded last. 

The entire startup is complex - attempt to use a generic interface obscures the startup logic
and prevents doing things in parallel, for example the wait for signing a cert can be done in background.